t/apicast-policy-headers.t

```diff
@@ -0,0 +1,355 @@
+                    { "op": "set", "header": "Header-To-Delete", "value": "" }
+                  ]
+              }
+          }
+        ],
+        "api_backend": "http://test:$TEST_NGINX_SERVER_PORT/",
+        "proxy_rules": [
+          { "pattern": "/", "http_method": "GET", "metric_system_name": "hits", "delta": 2 }
+        ]
+      }
+    }
+  ]
+}
+--- backend
+  location /transactions/authrep.xml {
+    content_by_lua_block {
+      local expected = "service_token=token-value&service_id=42&usage%5Bhits%5D=2&user_key=value"
+      local args = ngx.var.args
+      if args == expected then
+        ngx.exit(200)
+      else
+        ngx.log(ngx.ERR, expected, ' did not match: ', args)
+        ngx.exit(403)
+      end
+    }
+  }
+--- upstream
+  location / {
+     content_by_lua_block {
+       if ngx.req.get_headers()['New-Header'] ~= 'config_value_nh' then
+         ngx.log(ngx.ERR, "The header 'New-Header' does not have the correct value")
+       elseif ngx.req.get_headers()['Existing-Header'] ~= 'config_value_eh' then
+         ngx.log(ngx.ERR, "The header 'Existing-Header' does not have the correct value")
+       elseif ngx.req.get_headers()['Header-To-Delete'] then
+         ngx.log(ngx.ERR, "Received unexpected 'Header-To-Delete' header")
+       else
+         ngx.say('yay, api backend');
+       end
+     }
+  }
+--- request
+GET /?user_key=value
+--- more_headers
+Existing-Header: request_value_eh1, request_value_eh2
+Header-To-Delete: request_value_htd
+--- response_body
+yay, api backend
+--- response_headers
+New-Header:
+Existing-Header:
+Header-To-Delete:
+--- error_code: 200
+--- no_error_log
+[error]
+
+=== TEST 2: 'add' operation in request headers
+We test 3 things:
+1) Add op with a header that does not exist creates it with the given value.
+2) Add op with a header that exists, appends the given value to the existing one.
+3) Check that the headers are received in the upstream but not in the response
+   of the original request.
+--- configuration
+{
+  "services": [
+    {
+      "id": 42,
+      "backend_version":  1,
+      "backend_authentication_type": "service_token",
+      "backend_authentication_value": "token-value",
+      "proxy": {
+        "policy_chain": [
+          { "name": "apicast.policy.apicast" },
+          {
+            "name": "apicast.policy.headers",
+            "configuration":
+              {
+                "request":
+                  [
+                    { "op": "add", "header": "New-Header", "value": "config_value_nh" },
+                    { "op": "add", "header": "Existing-Header", "value": "config_value_eh" }
+                  ]
+              }
+          }
+        ],
+        "api_backend": "http://test:$TEST_NGINX_SERVER_PORT/",
+        "proxy_rules": [
+          { "pattern": "/", "http_method": "GET", "metric_system_name": "hits", "delta": 2 }
+        ]
+      }
+    }
+  ]
+}
+--- backend
+  location /transactions/authrep.xml {
```

Comment:
Do we want to have this in each policy test? I'd say it might be too much and actually testing the apicast policy.

Reply:
I was thinking the same. Would be good to have a way of declaring a 'default backend' so we don't copy paste the same thing over and over :)

Reply:
I agree that this is too verbose, but at the same time, I think it's important to check that adding this policy does not break the normal auth flow.

Reply:
Yes, but we could verify it with the phase logger policy? Just add it after and see it works.

And we can provide default backend in the blackbox test base. If no backend is defined we can add there whatever.

Reply:
I'd rather address this in a separate PR, as this is a problem that we have in multiple test files.

Reply:
I'd love get rid of those. Or at least have it in just one test. Because this is making plenty of noise that is not helping the test readability.

Reply:
@davidor missed your previous comment. Yes we can address it in different PR. 👍

Reply:
Also when dealing with this we could consider making api_backend and backend endpoint optional - that could disable the call and result in smaller more focused configuration file. 
I guess when writing these files by hand we want them to be as small and focused as possible - as our tests.